README.md
[![NPM version](https://badge.fury.io/js/cdk-serverless-lamp.svg)](https://badge.fury.io/js/cdk-serverless-lamp)
[![PyPI version](https://badge.fury.io/py/cdk-serverless-lamp.svg)](https://badge.fury.io/py/cdk-serverless-lamp)
![Build](https://github.com/aws-samples/cdk-serverless-lamp/workflows/Build/badge.svg)

# Welcome to cdk-serverless-lamp

`cdk-serverless-lamp` is a JSII construct library for AWS CDK that allows you to deploy the [New Serverless LAMP Stack](https://aws.amazon.com/tw/blogs/compute/introducing-the-new-serverless-lamp-stack/) running PHP Laravel Apps by specifying the local `laravel` directory.

By deploying the `ServerlessLaravel` and `DatabaseCluster`, the following resources will be created:

1. Amazon API Gateway HTTP API
2. AWS Lambda custom runtime with [Bref runtime](https://bref.sh/docs/runtimes/) support
3. Amazon Aurora for MySQL database cluster with RDS proxy enabled

## Howto

Create a new Laravel project with AWS CDK

```sh
$ mkdir serverless-lamp && cd serverless-lamp
# create cdk and codebase directories for AWS CDK and Laravel project
$ mkdir cdk codebase
# create the new Laravel project with docker
$ docker run --rm -ti \
  --volume $PWD:/app \
  composer create-project --prefer-dist laravel/laravel ./codebase
# install bref/bref and bref/laravel-bridge in the vendor
$ cd codebase
$ docker run --rm -ti \
  --volume $PWD:/app \
  composer require bref/bref bref/laravel-bridge
# initialize the AWS CDK project
$ cd ../cdk
$ cdk init -l typescript
# install the cdk-severless-lamp npm module
$ yarn add cdk-serverless-lamp
```

Now your directories should look like this:

```
.
├── cdk
└── codebase
```
where `cdk` is for the AWS CDK and `codebase` for Laravel project.

# AWS CDK sample

Building your serverless Laravel with `ServerlessLaravel` construct:

Update `./cdk/lib/cdk-stack.ts`

```ts
import * as cdk from '@aws-cdk/core';
import * as path from 'path';
import { ServerlessLaravel } from 'cdk-serverless-lamp';

export class CdkStack extends cdk.Stack {
  constructor(scope: cdk.Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    new ServerlessLaravel(this, 'ServerlessLaravel', {
      brefLayerVersion: 'arn:aws:lambda:us-east-1:209497400698:layer:php-74-fpm:12',
      laravelPath: path.join(__dirname, '../../codebase'),
    });
  }
}
```

deploy the CDK stack:

```sh
# see the difference before the deployment
$ cdk diff
# deploy it
$ cdk deploy
```


On deploy complete, the API Gateway URL will be returned in the Output. Click the URL and you will see the Laravel landing page:

![laravel-welcome](./images/laravel.png)


## Amazon Aurora support

Use `DatabaseCluster` to create the your database cluster:

```ts
import * as cdk from '@aws-cdk/core';
import { InstanceType, Vpc } from '@aws-cdk/aws-ec2';
import * as path from 'path';
import { ServerlessLaravel, DatabaseCluster } from 'cdk-serverless-lamp';

export class CdkStack extends cdk.Stack {
  constructor(scope: cdk.Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    const vpc = new Vpc(this, 'Vpc',{ maxAzs: 3, natGateways: 1 } )

    // the DatabaseCluster sharing the same vpc with the ServerlessLaravel
    const db = new DatabaseCluster(this, 'DatabaseCluster', {
      vpc,
      instanceType: new InstanceType('t3.small'),
      rdsProxy: true,
    })

    // the ServerlessLaravel
    new ServerlessLaravel(this, 'ServerlessLaravel', {
      brefLayerVersion: 'arn:aws:lambda:us-east-1:209497400698:layer:php-74-fpm:12',
      laravelPath: path.join(__dirname, '../../codebase'),
      vpc, 
      databaseConfig: {
        writerEndpoint: db.rdsProxy!.endpoint,
      },
    });
  }
}
 ```


## Local Development

Create `docker-compose.yml` with the following content:

```docker-compose
version: "3.5"
services:
  web:
    image: bref/fpm-dev-gateway
    ports:
      - "8000:80"
    volumes:
      - ./laravel:/var/task
    depends_on:
      - php
    environment:
      HANDLER: public/index.php
  php:
    image: bref/php-74-fpm-dev
    volumes:
      - ./laravel:/var/task
```

and run this command `docker-compose up -d` and now you can access <http://localhost:8000>.

_(more information can be found in [bref documentation](https://bref.sh/docs/local-development.html))_


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [1.4.43](https://github.com/aws-samples/cdk-serverless-lamp/compare/v1.4.41...v1.4.43) (2021-02-02)


### Bug Fixes

* **core:** LambdaProxyIntegration ([501b7b3](https://github.com/aws-samples/cdk-serverless-lamp/commit/501b7b34307595b799ce9f541f2fb7879ff12f2c))

### 1.4.42 (2020-11-28)

### 1.4.41 (2020-10-19)

### 1.4.40 (2020-10-19)

### 1.4.39 (2020-10-19)

### 1.4.38 (2020-10-19)

### 1.4.37 (2020-10-18)

### 1.4.36 (2020-10-16)

### 1.4.35 (2020-10-15)

### 1.4.34 (2020-10-03)

### 1.4.33 (2020-09-29)

### 1.4.32 (2020-09-29)

### 1.4.31 (2020-09-29)

### 1.4.30 (2020-09-29)

### 1.4.29 (2020-09-29)

### 1.4.28 (2020-09-29)

### 1.4.27 (2020-09-29)

### 1.4.26 (2020-09-29)

### 1.4.25 (2020-09-29)

### 1.4.24 (2020-09-25)

### 1.4.23 (2020-09-22)

### [1.4.22](https://github.com/aws-samples/cdk-serverless-lamp/compare/v1.4.21...v1.4.22) (2020-09-11)

### [1.4.21](https://github.com/aws-samples/cdk-serverless-lamp/compare/v1.4.19...v1.4.21) (2020-09-09)

### 1.4.20 (2020-09-06)

### [1.4.19](https://github.com/aws-samples/cdk-serverless-lamp/compare/v1.4.18...v1.4.19) (2020-09-06)

### 1.4.18 (2020-09-03)

### [1.4.17](https://github.com/aws-samples/cdk-serverless-lamp/compare/v1.4.16...v1.4.17) (2020-09-03)

### [1.4.16](https://github.com/aws-samples/cdk-serverless-lamp/compare/v1.4.15...v1.4.16) (2020-09-01)

### [1.4.15](https://github.com/aws-samples/cdk-serverless-lamp/compare/v1.4.14...v1.4.15) (2020-09-01)

### 1.4.14 (2020-08-28)

### 1.4.13 (2020-08-28)

### 1.4.12 (2020-08-28)

### 1.4.11 (2020-08-28)

### [1.4.10](https://github.com/aws-samples/cdk-serverless-lamp/compare/v1.4.9...v1.4.10) (2020-08-28)

### [1.4.9](https://github.com/aws-samples/cdk-serverless-lamp/compare/v1.4.8...v1.4.9) (2020-08-28)

### [1.4.8](https://github.com/aws-samples/cdk-serverless-lamp/compare/v1.4.7...v1.4.8) (2020-08-27)

### [1.4.7](https://github.com/aws-samples/cdk-serverless-lamp/compare/v1.4.5...v1.4.7) (2020-08-27)

### 1.4.6 (2020-08-27)

### 1.4.5 (2020-08-24)

### [1.4.4](https://github.com/aws-samples/cdk-serverless-lamp/compare/v1.4.2...v1.4.4) (2020-08-24)

### [1.4.3](https://github.com/aws-samples/cdk-serverless-lamp/compare/v1.4.1...v1.4.3) (2020-08-24)

### 1.4.3 (2020-08-24)

### 1.4.2 (2020-08-19)

### [1.4.1](https://github.com/aws-samples/cdk-serverless-lamp/compare/v1.4.0...v1.4.1) (2020-08-17)

## 1.4.0 (2020-08-17)


### Features

* **core:** add ServerlessLaravel to create Laravel app ([cd1ecbf](https://github.com/pahud/cdk-serverless-lamp/commit/cd1ecbf9a541b116b19cab460696ea2b6cd1cda8))
* **database:** support Amazon RDS and Amazon RDS Proxy ([53cc3de](https://github.com/pahud/cdk-serverless-lamp/commit/53cc3de3721ba3aa87964fa96614658176a47517))
* **serverless-api:** support laravel framework with brefi ([3820d7b](https://github.com/pahud/cdk-serverless-lamp/commit/3820d7b0c6e40ce4cba3fbd73d0ceb87aee9c4ff))


### Bug Fixes

* **api:** do not create new Vpc if props.vpc undefinied ([e9719bd](https://github.com/pahud/cdk-serverless-lamp/commit/e9719bdce407d22773f9ec33c6ca2ceeb4cbfa40)), closes [#20](https://github.com/pahud/cdk-serverless-lamp/issues/20)
* **database:** instanceCapacity ([bf11ff4](https://github.com/pahud/cdk-serverless-lamp/commit/bf11ff472777f84f72773a7efc57362aa633f197))
* **rdsPoxy:** rdsProxy is not enabled by default ([acf513b](https://github.com/pahud/cdk-serverless-lamp/commit/acf513bab378565f4c047e4082e4b1022917e007)), closes [#29](https://github.com/pahud/cdk-serverless-lamp/issues/29)

### [1.3.14](https://github.com/pahud/cdk-serverless-lamp/compare/v1.3.13...v1.3.14) (2020-08-15)

### [1.3.13](https://github.com/pahud/cdk-serverless-lamp/compare/v1.3.12...v1.3.13) (2020-08-13)

### 1.3.12 (2020-08-11)


### Bug Fixes

* **rdsPoxy:** rdsProxy is not enabled by default ([3c91fbb](https://github.com/pahud/cdk-serverless-lamp/commit/3c91fbb7c371f1e58026fa2b732dc5080b4d0861)), closes [#29](https://github.com/pahud/cdk-serverless-lamp/issues/29)

### 1.3.11 (2020-08-10)

### 1.3.10 (2020-08-10)

### 1.3.9 (2020-08-08)

### [1.3.8](https://github.com/pahud/cdk-serverless-lamp/compare/v1.3.7...v1.3.8) (2020-08-08)

### 1.3.7 (2020-08-04)

### [1.3.6](https://github.com/pahud/cdk-serverless-lamp/compare/v1.3.5...v1.3.6) (2020-08-04)


### Bug Fixes

* **api:** do not create new Vpc if props.vpc undefinied ([612e93d](https://github.com/pahud/cdk-serverless-lamp/commit/612e93d7c0f709c10175d40bc31a2abf92091fb3)), closes [#20](https://github.com/pahud/cdk-serverless-lamp/issues/20)

### [1.3.5](https://github.com/pahud/cdk-serverless-lamp/compare/v1.3.4...v1.3.5) (2020-08-04)

### 1.3.4 (2020-08-04)

### [1.3.3](https://github.com/pahud/cdk-serverless-lamp/compare/v1.3.2...v1.3.3) (2020-08-04)

### 1.3.3 (2020-08-04)

### [1.3.2](https://github.com/pahud/cdk-serverless-lamp/compare/v1.3.1...v1.3.2) (2020-08-04)

### [1.3.1](https://github.com/pahud/cdk-serverless-lamp/compare/v1.3.0...v1.3.1) (2020-07-29)


### Bug Fixes

* **database:** instanceCapacity ([5615d23](https://github.com/pahud/cdk-serverless-lamp/commit/5615d236ed44205d6ec925b470e48903ad546170))

## [1.3.0](https://github.com/pahud/cdk-serverless-lamp/compare/v1.2.1...v1.3.0) (2020-07-29)


### Features

* **database:** support Amazon RDS and Amazon RDS Proxy ([7b4c8a0](https://github.com/pahud/cdk-serverless-lamp/commit/7b4c8a0629f403d1ca885becb4dd3f75a9035f02))

## [1.3.0](https://github.com/pahud/cdk-serverless-lamp/compare/v1.2.1...v1.3.0) (2020-07-29)


### Features

* **database:** support Amazon RDS and Amazon RDS Proxy ([7b4c8a0](https://github.com/pahud/cdk-serverless-lamp/commit/7b4c8a0629f403d1ca885becb4dd3f75a9035f02))

### [1.2.1](https://github.com/pahud/cdk-serverless-lamp/compare/v1.2.0...v1.2.1) (2020-07-27)


src/integ.default.ts
import * as path from 'path';
import { Vpc, InstanceType } from '@aws-cdk/aws-ec2';
import * as cdk from '@aws-cdk/core';
import { ServerlessLaravel, DatabaseCluster } from './index';

export class IntegTesting {
  readonly stack: cdk.Stack[];

  constructor() {
    const app = new cdk.App();
    const env = {
      region: process.env.CDK_DEFAULT_REGION,
      account: process.env.CDK_DEFAULT_ACCOUNT,
    };

    const stack = new cdk.Stack(app, 'testing-stack', { env });

    const vpc = new Vpc(stack, 'Vpc', { maxAzs: 3, natGateways: 1 });

    // the DatabaseCluster sharing the same vpc with the ServerlessLaravel
    const db = new DatabaseCluster(stack, 'DatabaseCluster', {
      vpc,
      instanceType: new InstanceType('t3.small'),
      rdsProxy: true,
      instanceCapacity: 1,
    });

    // the ServerlessLaravel
    new ServerlessLaravel(stack, 'ServerlessLaravel', {
      brefLayerVersion: 'arn:aws:lambda:ap-northeast-1:209497400698:layer:php-74-fpm:11',
      laravelPath: path.join(__dirname, '../codebase'),
      vpc,
      databaseConfig: {
        writerEndpoint: db.rdsProxy!.endpoint,
      },
    });

    new cdk.CfnOutput(stack, 'RDSProxyEndpoint', { value: db.rdsProxy!.endpoint });
    new cdk.CfnOutput(stack, 'DBMasterUser', { value: db.masterUser });
    new cdk.CfnOutput(stack, 'DBMasterPasswordSecret', { value: db.masterPassword.secretArn });

    this.stack = [stack];
  }
}

// run the integ testing
new IntegTesting();
